### PR TITLE
feat: create sa and add self-provisioner role 

### PR DESCRIPTION
**Pull request by dipak-pawar**
Fixes: #2

Note: I will be planning to add E2E in the next PR

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package client
+
+import (
+	"k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ClientInterface interface {
```
Comment (xcoulon) [inline pkg/client/client.go]: what about calling the interface `Client` and the impl `clientImpl` ? so you wouldn't expose the `Interface` suffix everywhere.

File: pkg/client/clusterrolebinding.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"context"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// CreateClusterRoleBinding creates the ClusterRoleBinding.
+func (c *Client) CreateClusterRoleBinding(crb *rbacv1.ClusterRoleBinding) error {
+	return c.Create(context.Background(), crb)
```
Comment (xcoulon) [inline pkg/client/clusterrolebinding.go]: are you calling the `Create` method on the same `Client` struct?

File: pkg/client/clusterrolebinding.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+// GetClusterRoleBinding returns the existing ClusteRoleBinding.
+func (c *Client) GetClusterRoleBinding(name string) (*rbacv1.ClusterRoleBinding, error) {
+	crb := &rbacv1.ClusterRoleBinding{}
+
+	if err := c.Get(context.Background(), types.NamespacedName{Name: name}, crb); err != nil {
+		return nil, err
```
Comment (xcoulon) [inline pkg/client/clusterrolebinding.go]: what about logging or wrapping the error with a contextual message?

File: pkg/client/secret.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"context"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// CreateSecret creates the Secret.
+func (c *Client) CreateSecret(s *v1.Secret) error {
+	return c.Create(context.Background(), s)
```
Comment (xcoulon) [inline pkg/client/secret.go]: same question as above

File: pkg/client/secret.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+}
+
+// GetSecret returns the existing Secret.
+func (c *Client) GetSecret(namespace, name string) (*v1.Secret, error) {
+	s := &v1.Secret{}
+	if err := c.Get(context.Background(), types.NamespacedName{Namespace: namespace, Name: name}, s); err != nil {
+		return nil, err
```
Comment (xcoulon) [inline pkg/client/secret.go]: same question about logging/wrapping the error

File: pkg/client/serviceaccount.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import (
+	"context"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// CreateServiceAccount creates the serviceAccount.
+func (c *Client) CreateServiceAccount(sa *v1.ServiceAccount) error {
+	return c.Create(context.Background(), sa)
```
Comment (xcoulon) [inline pkg/client/serviceaccount.go]: same question about the call to `Create`

File: pkg/client/serviceaccount.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+// GetServiceAccount returns the existing serviceAccount.
+func (c *Client) GetServiceAccount(namespace, name string) (*v1.ServiceAccount, error) {
+	sa := &v1.ServiceAccount{}
+	if err := c.Get(context.Background(), types.NamespacedName{Namespace: namespace, Name: name}, sa); err != nil {
```
Comment (xcoulon) [inline pkg/client/serviceaccount.go]: same question about the error logging/wrapping

File: pkg/client/client.go (hunk 1 above)
Comment (dipak-pawar) [reply pkg/client/client.go]: sure, good idea. Let me do that

File: pkg/client/clusterrolebinding.go (hunk 2 above)
Comment (dipak-pawar) [reply pkg/client/clusterrolebinding.go]: Yes, we can call `c.Client` a split client which reads (Get and List) from the Cache and writes (Create, Update, Delete) to the API server.

File: pkg/client/clusterrolebinding.go (hunk 3 above)
Comment (dipak-pawar) [reply pkg/client/clusterrolebinding.go]: controller runtime is logging this error when there is an error in reconcile. So I think if we log it there we'll get it two times which would be confusing.

see https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L217

File: pkg/client/serviceaccount.go (hunk 6 above)
Comment (dipak-pawar) [reply pkg/client/serviceaccount.go]: addressed in c9c2328

File: pkg/client/secret.go (hunk 4 above)
Comment (dipak-pawar) [reply pkg/client/secret.go]: addressed in c9c2328

File: pkg/client/clusterrolebinding.go (hunk 2 above)
Comment (dipak-pawar) [reply pkg/client/clusterrolebinding.go]: Addressed in c9c2328

File: pkg/controller/toolchainenabler/toolchainenabler_controller.go [hunk 8]
```diff
@@ -152,12 +152,7 @@ func (r *ReconcileToolChainEnabler) addClusterRoleToSA(cr *codereadyv1alpha1.Too
 	}
 	crb.SetName("system:toolchain-enabler:self-provisioner")
 
-	//// Set ToolChainEnabler instance as the owner and controller
-	//if err := controllerutil.SetControllerReference(cr, crb, r.scheme); err != nil {
-	//	return err
-	//}
-
-	_, err := r.kubeClientSet.GetClusterRoleBinding("toolchain-enabler-system:self-provisioner")
+		_, err := r.kubeClientSet.GetClusterRoleBinding("system:toolchain-enabler:self-provisioner")
```
Comment (alexeykazakov) [inline pkg/controller/toolchainenabler/toolchainenabler_controller.go]: looks like broken formatting?

File: pkg/client/client.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+type Client interface {
+	client.Client
+	Secret
+	ServiceAccount
+	ClusterRoleBinding
+}
+
+// Secret contains methods for manipulating Secrets
+type Secret interface {
+	CreateSecret(*v1.Secret) error
```
Comment (alexeykazakov) [inline pkg/client/client.go]: It's not used anywhere?

File: pkg/controller/toolchainenabler/toolchainenabler_controller.go [hunk 10]
```diff
@@ -21,10 +22,10 @@ import (
 
 var log = logf.Log.WithName("controller_toolchainenabler")
 
-/**
-* USER ACTION REQUIRED: This is a scaffold file intended for the user to modify with their own Controller
-* business logic.  Delete these comments after modifying this file.*
- */
+var (
```
Comment (alexeykazakov) [inline pkg/controller/toolchainenabler/toolchainenabler_controller.go]: Use const instead.

File: pkg/client/client.go (hunk 9 above)
Comment (dipak-pawar) [reply pkg/client/client.go]: ok, let me remove it. will add if needed in future

**Pull request merged**